device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTokenRefresh.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.Client
+{
+    using System;
+    using Microsoft.Azure.Devices.Client.Extensions;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Authentication method that uses a shared access signature token and allows for token refresh. 
+    /// </summary>
+    public abstract class DeviceAuthenticationWithTokenRefresh : IAuthenticationMethod
+    {
+        private string _deviceId;
+        private readonly int _timeToLiveBufferSeconds;
+        private SemaphoreSlim _lock = new SemaphoreSlim(1);
+        private string _token;
+        private DateTime _expiryTime;
+
+        public string DeviceId 
+        {
+            get
+            {
+                return _deviceId;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeviceAuthenticationWithTokenRefresh"/> class.
+        /// </summary>
+        /// <param name="deviceId">Device Identifier.</param>
+        public DeviceAuthenticationWithTokenRefresh(string deviceId, 
+            int timeToLiveBufferSeconds = 60) 
+        {
+            if (deviceId.IsNullOrWhiteSpace())
+            {
+                throw new ArgumentNullException(nameof(deviceId));
+            }
+
+            if (timeToLiveBufferSeconds < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeToLiveBufferSeconds));
+            }
+
+            _deviceId = deviceId;
+            _timeToLiveBufferSeconds = timeToLiveBufferSeconds;
+        }
+
+        /// <summary>
+        /// Gets a snapshot of the security token associated with the device. This call is thread-safe.
+        /// </summary>
+        public async Task<string> GetTokenAsync(string iotHubName)
+        {
+            if (!IsExpiring)
+            {
+                return _token;
+            }
+
+            await _lock.WaitAsync();
+
+            if (!IsExpiring)
```

Comment:
Why are we doing this twice? Is it to make sure that once we take application lock that it is still good?

Reply:
It's a double checked lock pattern (see https://en.wikipedia.org/wiki/Double-checked_locking).